pkg/storage/image.go

```diff
@@ -176,145 +178,106 @@ func (svc *imageService) makeRepoDigests(knownRepoDigests, tags []string, imageI
 	return imageDigest, repoDigests
 }
 
+func (svc *imageService) buildImageCacheItem(systemContext *types.SystemContext, ref types.ImageReference, image *storage.Image) (imageCacheItem, error) {
+	img, err := ref.NewImageSource(systemContext)
+	if err != nil {
+		return imageCacheItem{}, err
+	}
+	size := imageSize(img)
+	configDigest, err := imageConfigDigest(img, nil)
+	img.Close()
+	if err != nil {
+		return imageCacheItem{}, err
+	}
+	imageFull, err := ref.NewImage(systemContext)
+	if err != nil {
+		return imageCacheItem{}, err
+	}
+	defer imageFull.Close()
+	imageConfig, err := imageFull.OCIConfig()
+	if err != nil {
+		return imageCacheItem{}, err
+	}
+	return imageCacheItem{
+		user:         imageConfig.Config.User,
+		size:         size,
+		configDigest: configDigest,
+	}, nil
+}
+
+func (svc *imageService) appendCachedResult(systemContext *types.SystemContext, ref types.ImageReference, image *storage.Image, results []ImageResult, newImageCache imageCache) ([]ImageResult, error) {
+	var err error
+	svc.imageCacheLock.Lock()
+	cacheItem, ok := svc.imageCache[image.ID]
+	svc.imageCacheLock.Unlock()
+	if !ok {
+		cacheItem, err = svc.buildImageCacheItem(systemContext, ref, image)
+		if err != nil {
+			return results, err
+		}
+		if newImageCache == nil {
+			svc.imageCacheLock.Lock()
+			svc.imageCache[image.ID] = cacheItem
+			svc.imageCacheLock.Unlock()
+		} else {
+			newImageCache[image.ID] = cacheItem
+		}
+	} else if newImageCache != nil {
+		newImageCache[image.ID] = cacheItem
+	}
+
+	name, tags, digests := sortNamesByType(image.Names)
```

Comment:
Why caching some, but not the rest? shouldn't we cache all these computations too?

Reply:
An image's `Names` can change while the image is on the client if tags are added or removed from the image, and we need the naming information to be correct when we hand back the results to the CRI client, so we don't try to cache the names.